src/ota/mod.rs

```diff
@@ -104,3 +126,2053 @@ impl Default for DeployStatus {
+            OtaStatus::Init(id)
+            | OtaStatus::Acknowledged(id)
+            | OtaStatus::Downloading(id, _)
+            | OtaStatus::Deploying(id, _)
+            | OtaStatus::Deployed(id)
+            | OtaStatus::Rebooting(id)
+            | OtaStatus::Success(id)
+            | OtaStatus::Error(_, id) => Some(id.clone()),
+            OtaStatus::Failure(_, id) => id.clone(),
+        }
+    }
+}
+
+impl Display for OtaStatus {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            OtaStatus::Idle => write!(f, "Idle"),
+            OtaStatus::Init(req) => write!(f, "Init {req}"),
+            OtaStatus::NoPendingOta => write!(f, "NoPendingOta"),
+            OtaStatus::Acknowledged(req) => write!(f, "Acknowledged {req}"),
+            OtaStatus::Downloading(req, progress) => {
+                write!(f, "Downloading {req} progress {progress}")
+            }
+            OtaStatus::Deploying(req, progress) => write!(f, "Deploying {req} {progress}"),
+            OtaStatus::Deployed(req) => write!(f, "Deployed {req}"),
+            OtaStatus::Rebooting(req) => write!(f, "Rebooting {req}"),
+            OtaStatus::Rebooted => write!(f, "Rebooted"),
+            OtaStatus::Success(req) => write!(f, "Success {req}"),
+            OtaStatus::Error(err, req) => write!(f, "Error {req}: {err}"),
+            OtaStatus::Failure(err, req) => {
+                write!(f, "Failure")?;
+
+                if let Some(req) = req {
+                    write!(f, " {req}")?;
+                }
+
+                write!(f, ": {err}")
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct OtaId {
+    pub uuid: Uuid,
+    pub url: String,
+}
+
+impl From<OtaRequest> for OtaId {
+    fn from(value: OtaRequest) -> Self {
+        Self {
+            uuid: value.uuid.0,
+            url: value.url,
+        }
+    }
+}
+
+impl Display for OtaId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.uuid)
+    }
+}
+
+impl OtaStatus {
+    pub fn ota_request(&self) -> Option<&OtaId> {
+        match self {
+            OtaStatus::Acknowledged(ota_request)
+            | OtaStatus::Downloading(ota_request, _)
+            | OtaStatus::Deploying(ota_request, _)
+            | OtaStatus::Deployed(ota_request)
+            | OtaStatus::Rebooting(ota_request)
+            | OtaStatus::Success(ota_request)
+            | OtaStatus::Error(_, ota_request) => Some(ota_request),
+            OtaStatus::Failure(_, ota_request) => ota_request.as_ref(),
+            _ => None,
+        }
+    }
+}
+
+/// Provides OTA resource accessibility only by talking with it.
+pub struct Ota<T, U>
+where
+    T: SystemUpdate,
+    U: StateRepository<PersistentState>,
+{
+    pub system_update: T,
+    pub state_repository: U,
+    pub download_file_path: PathBuf,
+    pub ota_status: OtaStatus,
+    pub flag: OtaInProgress,
```

Comment:
I'm not sure about introducing another flag to obtain information that you can already get from status, and keep in sync with the status. Perhaps would  be better to add a function `is_in_progress` that returns a boolean from the `ota_status`. What do you think?

Reply:
The flag/status need to be shared with the OtaHandle, this can be achieved by wrapping the status in a Mutex, but the only information we want to know is if the OtaStatus is not Idle.

I think this is better achieved with the AtomicBool (the OtaInProgress struct), but both are viable options.

Just tell me what you prefer